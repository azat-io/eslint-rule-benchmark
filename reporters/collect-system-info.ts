import { createRequire } from 'node:module'
import fs from 'node:fs/promises'
import os from 'node:os'

/** System information collected during benchmark execution. */
export interface SystemInfo {
  /** Total system memory in gigabytes. */
  totalMemoryGb: number

  /** ESLint version used during benchmarking. */
  eslintVersion: string

  /** Node.js version (e.g., "v20.11.0"). */
  nodeVersion: string

  /** CPU base frequency in MHz. */
  cpuSpeedMHz: number

  /** Operating system release version (e.g., "22.6.0", "10.0.22621"). */
  osRelease: string

  /** V8 JavaScript engine version. */
  v8Version: string

  /** Operating system platform (e.g., "darwin", "linux", "win32"). */
  platform: string

  /** Number of CPU cores. */
  cpuCount: number

  /** CPU model (e.g., "Intel(R) Core(TM) i7-12700K CPU @ 3.60GHz"). */
  cpuModel: string

  /** CPU architecture (e.g., "x64", "arm64"). */
  arch: string
}

/**
 * Collects system information that may affect benchmark results.
 *
 * @returns Object containing system specifications and versions.
 */
export async function collectSystemInfo(): Promise<SystemInfo> {
  let require = createRequire(import.meta.url)
  let eslintPackagePath = require.resolve('eslint/package.json')
  let eslintPackageContent = await fs.readFile(eslintPackagePath, 'utf8')
  let eslintPackage = JSON.parse(eslintPackageContent) as { version: string }
  let eslintVersion = eslintPackage.version

  let totalMemoryBytes = os.totalmem()
  let totalMemoryGb = Math.round(totalMemoryBytes / (1024 * 1024 * 1024))

  let cpus = os.cpus()
  let [firstCpu] = cpus

  return {
    v8Version: process.versions.v8,
    nodeVersion: process.version,
    cpuSpeedMHz: firstCpu!.speed,
    cpuModel: firstCpu!.model,
    platform: os.platform(),
    osRelease: os.release(),
    cpuCount: cpus.length,
    arch: os.arch(),
    totalMemoryGb,
    eslintVersion,
  }
}
