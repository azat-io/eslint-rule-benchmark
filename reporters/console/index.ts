import type { SingleRuleResult } from '../../runners/run-single-rule'
import type { BenchmarkConfig } from '../../types/benchmark-config'

import { printSummary } from './print-summary'

/**
 * Creates a console reporter for benchmark results.
 *
 * @returns {(result: SingleRuleResult, config: BenchmarkConfig) => void} A
 *   function that reports benchmark results to the console.
 */
export let createConsoleReporter =
  (): ((result: SingleRuleResult, config: BenchmarkConfig) => void) =>
  (result: SingleRuleResult, config: BenchmarkConfig): void => {
    printSummary(result, {
      slowThresholdMs: config.timeout / 100,
    })
  }

/**
 * Console reporter factory function.
 *
 * @returns {(result: SingleRuleResult, config: BenchmarkConfig) => void}
 *   Reporter function.
 */
export let consoleReporter = (): ((
  result: SingleRuleResult,
  config: BenchmarkConfig,
) => void) => createConsoleReporter()
