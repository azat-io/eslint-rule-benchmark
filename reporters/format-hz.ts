import { formatNumber } from './format-number'

/**
 * Formats a given number of operations per second (Hz) into a string
 * representation with thousands separators.
 *
 * @param {number} hz - The number of operations per second (Hz) to format.
 * @returns {string} A string representation of the formatted Hz.
 */
export let formatHz = (hz?: number): string => {
  if (typeof hz !== 'number' || !Number.isFinite(hz)) {
    return 'N/A'
  }

  let rounded = Math.round(hz)
  let formatted = formatNumber(rounded)
  return `${formatted} ops/sec`
}
