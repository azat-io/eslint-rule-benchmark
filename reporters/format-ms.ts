import { formatNumber } from './format-number'

/**
 * Formats a given number of milliseconds into a string representation with
 * thousands separators.
 *
 * @param {number} milliseconds - The number of milliseconds to format.
 * @returns {string} A string representation of the formatted milliseconds.
 */
export function formatMs(milliseconds?: number): string {
  if (typeof milliseconds !== 'number' || !Number.isFinite(milliseconds)) {
    return 'N/A'
  }

  let fixed = milliseconds.toFixed(3)
  let [integerPart, decimalPart] = fixed.split('.')
  let formattedInteger = formatNumber(Number(integerPart))
  return `${formattedInteger}.${decimalPart} ms`
}
