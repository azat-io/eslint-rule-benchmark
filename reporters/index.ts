import type {
  SingleRuleResult,
  BenchmarkConfig,
  ReporterOptions,
} from '../types/benchmark-config'

import { markdownReporter } from './markdown'
import { consoleReporter } from './console'
import { jsonReporter } from './json'

/** Registry of available reporters. */
let reporters = {
  markdown: markdownReporter,
  console: consoleReporter,
  json: jsonReporter,
}

/** Reporter function type. */
type ReporterFunction = (
  result: SingleRuleResult,
  config: BenchmarkConfig,
) => void

/**
 * Creates a reporter based on the specified options.
 *
 * @param {ReporterOptions} options - Reporter configuration options.
 * @returns {ReporterFunction} A configured reporter function.
 */
let createReporter = (options: ReporterOptions): ReporterFunction => {
  let { format = 'console' } = options

  if (!(format in reporters)) {
    throw new Error(
      `Unknown reporter format "${format}". Available formats: ${new Intl.ListFormat(
        'en',
        {
          type: 'conjunction',
          style: 'long',
        },
      ).format(Object.keys(reporters))}`,
    )
  }

  return reporters[format](options)
}

/**
 * Runs all configured reporters for the benchmark result.
 *
 * @param {SingleRuleResult} result - The benchmark result to report.
 * @param {BenchmarkConfig} config - The benchmark configuration.
 * @returns {void}
 */
export let runReporters = (
  result: SingleRuleResult,
  config: BenchmarkConfig,
): void => {
  for (let reporterOptions of config.reporters) {
    try {
      let reporter = createReporter(reporterOptions)
      reporter(result, config)
    } catch (error) {
      let errorValue = error as Error
      console.error(
        `Error in reporter "${reporterOptions.format}": ${errorValue.message}`,
      )
    }
  }
}
