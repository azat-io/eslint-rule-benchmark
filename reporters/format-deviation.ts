import { formatMs } from './format-ms'

/**
 * Formats a deviation value in milliseconds for display.
 *
 * @param deviation - The deviation value in milliseconds.
 * @returns A formatted string representing the deviation.
 */
export function formatDeviation(deviation: number): string {
  if (typeof deviation !== 'number' || !Number.isFinite(deviation)) {
    return 'N/A'
  }
  return `Â±${formatMs(deviation)}`
}
