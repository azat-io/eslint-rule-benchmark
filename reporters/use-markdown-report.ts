import type { UserBenchmarkConfig } from '../types/user-benchmark-config'
import type { TestSpecResult } from '../types/benchmark-config'

import { formatNumber } from './format-number'
import { formatMs } from './format-ms'
import { formatHz } from './format-hz'

/**
 * Creates a markdown reporter for aggregated benchmark results.
 *
 * @param {TestSpecResult[]} results - An array of results for all test
 *   specifications.
 * @param {UserBenchmarkConfig} _userConfig - The user's benchmark
 *   configuration.
 * @returns {string} Formatted markdown report as a string.
 */
export let useMarkdownReport = (
  results: TestSpecResult[],
  _userConfig?: UserBenchmarkConfig,
): string => {
  let outputLines: string[] = []

  if (results.length === 0) {
    return 'No benchmark results available.'
  }

  outputLines.push('# ESLint Rule Benchmark Report')

  for (let i = 0; i < results.length; i++) {
    let testSpecResult = results[i]!

    outputLines.push('', `## ${testSpecResult.name}`, '')

    if (testSpecResult.testCaseResults.length === 0) {
      outputLines.push(
        'No test cases found or all failed for this specification.',
      )
      continue
    }

    let tableRows: string[] = []

    tableRows.push(
      '| Sample | Ops/sec | Avg Time | Median | Min | Max | StdDev | Samples |',
      '| ------ | ------- | -------- | ------ | --- | --- | ------ | ------- |',
    )

    for (let testCaseResult of testSpecResult.testCaseResults) {
      if (testCaseResult.samplesResults.length === 0) {
        tableRows.push(
          `| No samples | N/A | N/A | N/A | N/A | N/A | N/A | N/A |`,
        )
        continue
      }

      for (let sampleResult of testCaseResult.samplesResults) {
        let sampleName = sampleResult.name.replace(
          `${testCaseResult.name} on `,
          '',
        )
        let rowData = [
          sampleName,
          formatHz(sampleResult.metrics.hz),
          formatMs(sampleResult.metrics.mean),
          formatMs(sampleResult.metrics.median),
          formatMs(sampleResult.metrics.min),
          formatMs(sampleResult.metrics.max),
          formatMs(sampleResult.metrics.stdDev),
          formatNumber(sampleResult.metrics.sampleCount),
        ]

        tableRows.push(`| ${rowData.join(' | ')} |`)
      }
    }

    outputLines.push(...tableRows)

    if (i < results.length - 1) {
      outputLines.push('')
    }
  }

  return outputLines.join('\n')
}
