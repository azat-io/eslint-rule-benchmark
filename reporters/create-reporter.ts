import type { TestSpecResult, ReporterFormat } from '../types/benchmark-config'
import type { UserBenchmarkConfig } from '../types/user-benchmark-config'

import { useMarkdownReport } from './use-markdown-report'
import { useConsoleReport } from './use-console-report'
import { useJsonReport } from './use-json-report'

/** Registry of available reporters. */
let reporters: Record<
  ReporterFormat,
  (results: TestSpecResult[], config: UserBenchmarkConfig) => string
> = {
  markdown: useMarkdownReport,
  console: useConsoleReport,
  json: useJsonReport,
}

/**
 * Creates a reporter based on the specified options.
 *
 * @param {TestSpecResult[]} results - An array of results for all test
 *   specifications.
 * @param {UserBenchmarkConfig} userConfig - The user's benchmark configuration.
 * @param {ReporterFormat} format - The format of the report.
 * @returns {string} The formatted report as a string.
 */
export let createReporter = (
  results: TestSpecResult[],
  userConfig: UserBenchmarkConfig,
  format: ReporterFormat = 'console',
): string => {
  if (!(format in reporters)) {
    throw new Error(
      `Unknown reporter format "${format}". Available formats: ${new Intl.ListFormat(
        'en',
        {
          type: 'conjunction',
          style: 'long',
        },
      ).format(Object.keys(reporters))}`,
    )
  }

  return reporters[format](results, userConfig)
}
