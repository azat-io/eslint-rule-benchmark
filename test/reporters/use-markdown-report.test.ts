import { describe, expect, it, vi } from 'vitest'

import type {
  BenchmarkConfig,
  TestSpecResult,
  TestCaseResult,
} from '../../types/benchmark-config'
import type { ProcessedBenchmarkTask } from '../../core/benchmark/run-benchmark'
import type { UserBenchmarkConfig } from '../../types/user-benchmark-config'
import type { BenchmarkMetrics } from '../../types/benchmark-metrics'
import type { RuleConfig } from '../../types/test-case'

import { useMarkdownReport } from '../../reporters/use-markdown-report'

vi.mock('../../reporters/collect-system-info', () => ({
  collectSystemInfo: () => ({
    cpuModel: 'Intel(R) Core(TM) i7-12700K CPU @ 3.60GHz',
    v8Version: '11.3.244.8-node.20',
    osRelease: '6.2.0-39-generic',
    nodeVersion: 'v20.11.0',
    eslintVersion: '9.27.0',
    platform: 'linux',
    cpuSpeedMHz: 3600,
    totalMemoryGb: 32,
    arch: 'x64',
    cpuCount: 8,
  }),
}))

let createMockMetrics = (
  overrides: Partial<BenchmarkMetrics> = {},
): BenchmarkMetrics => ({
  sampleCount: 10,
  period: 0.001,
  stdDev: 0.05,
  median: 0.9,
  p75: 1.05,
  min: 0.8,
  max: 1.2,
  p99: 1.1,
  hz: 1000,
  mean: 1,
  ...overrides,
})

let createMockProcessedTask = (
  name: string,
  metricOverrides: Partial<BenchmarkMetrics> = {},
): ProcessedBenchmarkTask => ({
  metrics: createMockMetrics(metricOverrides),
  name,
})

let createMockRuleConfig = (
  ruleId: string = 'test-rule',
  path?: string,
  options?: unknown[],
): RuleConfig => ({
  options: options ?? undefined,
  severity: 2,
  ruleId,
  path,
})

interface CreateMockTestCaseResultParameters {
  samplesResults?: ProcessedBenchmarkTask[]
  ruleConfig?: RuleConfig
  description?: string
  name?: string
  id?: string
}

let createMockTestCaseResult = (
  parameters: CreateMockTestCaseResultParameters = {},
): TestCaseResult => {
  let name = parameters.name ?? 'Test Case 1'
  let id = parameters.id ?? 'tc-1'
  let ruleConfig = parameters.ruleConfig ?? createMockRuleConfig()
  let samplesResults = parameters.samplesResults ?? [
    createMockProcessedTask(`${name} on sampleA.js`),
  ]
  let { description } = parameters

  return {
    rule: ruleConfig,
    samplesResults,
    description,
    name,
    id,
  }
}

let createMockBenchmarkConfig = (
  overrides: Partial<
    Omit<BenchmarkConfig, 'baselinePath' | 'reporters' | 'name'>
  > = {},
): Omit<BenchmarkConfig, 'baselinePath' | 'reporters' | 'name'> => ({
  warmup: { enabled: true, iterations: 3 },
  iterations: 10,
  timeout: 5000,
  ...overrides,
})

interface CreateMockTestSpecResultParameters {
  benchmarkConfigOverrides?: Partial<
    Omit<BenchmarkConfig, 'baselinePath' | 'reporters' | 'name'>
  >
  testCaseResults?: TestCaseResult[]
  rulePath?: string
  ruleId?: string
  name?: string
}

let createMockTestSpecResult = (
  parameters: CreateMockTestSpecResultParameters = {},
): TestSpecResult => {
  let {
    ruleId = 'test-spec-rule',
    benchmarkConfigOverrides,
    name = 'Test Spec 1',
    rulePath,
  } = parameters
  let testCaseResults = parameters.testCaseResults ?? [
    createMockTestCaseResult({}),
  ]
  let benchmarkConfig = createMockBenchmarkConfig(benchmarkConfigOverrides)

  return {
    benchmarkConfig,
    testCaseResults,
    rulePath,
    ruleId,
    name,
  }
}

let createMockUserConfig = (
  overrides: Partial<UserBenchmarkConfig> = {},
): UserBenchmarkConfig => ({
  tests: [],
  ...overrides,
})

describe('useMarkdownReport', () => {
  it('returns complete markdown report for valid benchmark results', async () => {
    let sample1 = createMockProcessedTask(
      'Test Spec 1 - Test Case 1 on sampleA.js',
      { median: 0.9, hz: 1000, mean: 1 },
    )
    let sample2 = createMockProcessedTask(
      'Test Spec 1 - Test Case 1 on sampleB.js',
      { median: 1.9, mean: 2, hz: 500 },
    )

    let ruleConfig = createMockRuleConfig('my-rule', 'path/to/my-rule.js')
    let testCase1 = createMockTestCaseResult({
      samplesResults: [sample1, sample2],
      description: 'Description for TC1',
      name: 'Test Spec 1 - Test Case 1',
      id: 'ts1-tc1',
      ruleConfig,
    })

    let testSpec1 = createMockTestSpecResult({
      benchmarkConfigOverrides: { iterations: 100, timeout: 3000 },
      rulePath: 'path/to/my-rule.js',
      testCaseResults: [testCase1],
      name: 'My Rule Benchmarks',
      ruleId: 'my-rule',
    })

    let mockTestSpecResults: TestSpecResult[] = [testSpec1]
    let mockUserCfg = createMockUserConfig()

    let markdownOutput = await useMarkdownReport(
      mockTestSpecResults,
      mockUserCfg,
    )

    expect(markdownOutput).toContain('# ESLint Rule Benchmark Report')
    expect(markdownOutput).toContain('## My Rule Benchmarks')
    expect(markdownOutput).toContain(
      '| Sample | Ops/sec | Avg Time | Median | Min | Max | StdDev | Samples |',
    )
    expect(markdownOutput).toContain(
      '| ------ | ------- | -------- | ------ | --- | --- | ------ | ------- |',
    )
    expect(markdownOutput).toContain(
      '| sampleA.js | 1,000 ops/sec | 1.000 ms | 0.900 ms | 0.800 ms | 1.200 ms | ±0.050 ms | 10 |',
    )
    expect(markdownOutput).toContain(
      '| sampleB.js | 500 ops/sec | 2.000 ms | 1.900 ms | 0.800 ms | 1.200 ms | ±0.050 ms | 10 |',
    )
    expect(markdownOutput).not.toContain('Test Specification:')
    expect(markdownOutput).not.toContain('Rule ID:')
    expect(markdownOutput).not.toContain('### Test Case:')
    expect(markdownOutput).not.toContain('### Sample:')
    expect(markdownOutput).not.toContain('Performance Summary')
    expect(markdownOutput).not.toContain('Benchmark Configuration')
  })

  it('handles multiple test specifications', async () => {
    let testSpec1 = createMockTestSpecResult({
      testCaseResults: [
        createMockTestCaseResult({
          samplesResults: [
            createMockProcessedTask('Base case on base-case.ts', {
              sampleCount: 37011,
              median: 0.055,
              stdDev: 0.002,
              mean: 0.056,
              min: 0.052,
              max: 0.063,
              hz: 17896,
            }),
          ],
          name: 'Base case',
        }),
        createMockTestCaseResult({
          samplesResults: [
            createMockProcessedTask('Complex case on complex-case.ts', {
              sampleCount: 32886,
              median: 0.055,
              stdDev: 0.002,
              mean: 0.056,
              min: 0.053,
              max: 0.062,
              hz: 17973,
            }),
          ],
          name: 'Complex case',
        }),
      ],
      rulePath: '../rules/no-negated-conjunction.ts',
      name: 'Rule: no-negated-conjunction',
      ruleId: 'no-negated-conjunction',
    })

    let testSpec2 = createMockTestSpecResult({
      testCaseResults: [
        createMockTestCaseResult({
          samplesResults: [
            createMockProcessedTask('Base case on base-case.ts', {
              sampleCount: 34813,
              median: 0.055,
              stdDev: 0.002,
              mean: 0.056,
              min: 0.053,
              max: 0.063,
              hz: 17941,
            }),
          ],
          name: 'Base case',
        }),
      ],
      rulePath: '../rules/no-negated-disjunction.ts',
      name: 'Rule: no-negated-disjunction',
      ruleId: 'no-negated-disjunction',
    })

    let mockTestSpecResults: TestSpecResult[] = [testSpec1, testSpec2]
    let mockUserCfg = createMockUserConfig()

    let markdownOutput = await useMarkdownReport(
      mockTestSpecResults,
      mockUserCfg,
    )

    expect(markdownOutput).toMatchSnapshot()
  })

  it('handles test specification with no test cases', async () => {
    let testSpec = createMockTestSpecResult({
      name: 'Empty Test Specification',
      testCaseResults: [],
    })

    let mockTestSpecResults: TestSpecResult[] = [testSpec]
    let mockUserCfg = createMockUserConfig()

    let markdownOutput = await useMarkdownReport(
      mockTestSpecResults,
      mockUserCfg,
    )

    expect(markdownOutput).toContain('## Empty Test Specification')
    expect(markdownOutput).toContain(
      'No test cases found or all failed for this specification.',
    )
  })

  it('handles test case with no samples', async () => {
    let testCase = createMockTestCaseResult({
      name: 'Test Case with No Samples',
      samplesResults: [],
    })

    let testSpec = createMockTestSpecResult({
      name: 'Test Spec with Empty Test Case',
      testCaseResults: [testCase],
    })

    let mockTestSpecResults: TestSpecResult[] = [testSpec]
    let mockUserCfg = createMockUserConfig()

    let markdownOutput = await useMarkdownReport(
      mockTestSpecResults,
      mockUserCfg,
    )

    expect(markdownOutput).toContain(
      '| No samples | N/A | N/A | N/A | N/A | N/A | N/A | N/A |',
    )
  })

  it('handles empty results array', async () => {
    let mockTestSpecResults: TestSpecResult[] = []
    let mockUserCfg = createMockUserConfig()

    let markdownOutput = await useMarkdownReport(
      mockTestSpecResults,
      mockUserCfg,
    )

    expect(markdownOutput).toBe('No benchmark results available.')
  })

  it('formats metrics with N/A for invalid numbers', async () => {
    let sample = createMockProcessedTask('Invalid metrics sample', {
      stdDev: undefined,
      median: Infinity,
      mean: undefined,
      min: -Infinity,
      sampleCount: 0,
      hz: Number.NaN,
      max: undefined,
    })

    let testCase = createMockTestCaseResult({
      name: 'Test Case with Invalid Metrics',
      samplesResults: [sample],
    })

    let testSpec = createMockTestSpecResult({
      name: 'Test Spec with Invalid Metrics',
      testCaseResults: [testCase],
    })

    let mockTestSpecResults: TestSpecResult[] = [testSpec]
    let mockUserCfg = createMockUserConfig()

    let markdownOutput = await useMarkdownReport(
      mockTestSpecResults,
      mockUserCfg,
    )

    expect(markdownOutput).toContain(
      '| Invalid metrics sample | N/A | N/A | N/A | N/A | N/A | N/A | 0 |',
    )
  })
})
