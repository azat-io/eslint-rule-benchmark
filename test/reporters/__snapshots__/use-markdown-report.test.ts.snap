// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useMarkdownReport > handles multiple test specifications 1`] = `
"# ESLint Rule Benchmark Report

## Rule: no-negated-conjunction

| Sample | Ops/sec | Avg Time | Median | Min | Max | StdDev | Samples |
| ------ | ------- | -------- | ------ | --- | --- | ------ | ------- |
| base-case.ts | 17,896 ops/sec | 0.056 ms | 0.055 ms | 0.052 ms | 0.063 ms | 0.002 ms | 37,011 |
| complex-case.ts | 17,973 ops/sec | 0.056 ms | 0.055 ms | 0.053 ms | 0.062 ms | 0.002 ms | 32,886 |


## Rule: no-negated-disjunction

| Sample | Ops/sec | Avg Time | Median | Min | Max | StdDev | Samples |
| ------ | ------- | -------- | ------ | --- | --- | ------ | ------- |
| base-case.ts | 17,941 ops/sec | 0.056 ms | 0.055 ms | 0.053 ms | 0.063 ms | 0.002 ms | 34,813 |

## System Information

**Runtime:** Node.js v20.11.0, V8 11.3.244.8-node.20, ESLint 9.27.0

**Platform:** linux x64 (6.2.0-39-generic)

**Hardware:** Intel(R) Core(TM) i7-12700K CPU @ 3.60GHz (8 cores, 3600 MHz), 32 GB RAM"
`;
